lib/internal/url.js

```diff
@@ -0,0 +1,328 @@
+'use strict';
+
+function getPunycode() {
+  try {
+    return process.binding('icu');
+  } catch (err) {
+    return require('punycode');
+  }
+}
+const punycode = getPunycode();
+const binding = process.binding('url');
+const context = Symbol('context');
+const cannotBeBase = Symbol('cannot-be-base');
+
+function StorageObject() {}
+StorageObject.prototype = Object.create(null);
+
+class URL {
+  constructor(input, base) {
```

Comment:
This is not a good implementation of the IDL. input and base do not need to be strings (or URLs).

Instead, the first few lines of this should be something like

```js
if (input === undefined) {
  throw new TypeError("first argument is required");
}
input = String(input);

if (base !== undefined) {
  base = String(base);
}
```

(Actually those `String(input)`s should actually be `toUSVString(input)`... but that's a separate discussion.)

The same applies to all the setters, which should just do `scheme = String(scheme)`, not throw a TypeError.

These aspects of the API will not be tested in the web platform tests since they are, in typical implementations, auto-generated from Web IDL.